File: python/src/securitycontext/schema.py
from __future__ import annotations

import datetime as dt
import hashlib
import re

SCHEMA_VERSION = 2
FINGERPRINT_VERSION = 2
PRODUCT = "SecurityContext"


def sink_fields(rule, role, function, location):
    aliases = {"template": "sql_template", "shell": "shell_script", "shell_command": "shell_script", "argv": "argument", "ordinary_argument": "argument", "unknown_target": "destination_unknown", "request_path": "path_or_query", "request_query": "path_or_query", "path_query": "path_or_query"}
    sink = {"function": function, "role": aliases.get(role, role), "location": location, "operation": "", "path_role": "", "input_part": ""}
    if rule == "http_request_input":
        sink["input_part"] = "path" if role == "request_path" else "query" if role == "request_query" else "path_or_query"
    if rule == "path_traversal":
        sink["role"] = "file_path"
        name = function.lower()
        operation = "unknown"
        if re.search(r"copy|\.cp(?:sync)?$|\.link(?:sync)?$", name):
            operation = "copy"
        elif re.search(r"rename|\.move$", name):
            operation = "rename"
        elif re.search(r"delete|unlink|rmdir|\.rm(?:sync)?$", name):
            operation = "delete"
        elif role in ("read", "write", "delete", "rename"):
            operation = role
        elif role == "source" or re.search(r"inputstream|reader|directorystream|\.read", name):
            operation = "read"
        elif role == "target" or re.search(r"outputstream|writer|\.write", name):
            operation = "write"
        sink["operation"] = operation
        sink["path_role"] = "unknown"
        if role in ("source", "read"):
            sink["path_role"] = "source"
        elif role in ("target", "write", "delete", "destination_path"):
            sink["path_role"] = "target"
        elif role == "file_path" and operation in ("copy", "rename", "read"):
            sink["path_role"] = "source"
        elif role == "file_path" and operation in ("write", "delete"):
            sink["path_role"] = "target"
    return sink


def finding_fingerprint(application_id, language, rule, sink, signatures):
    result = hashlib.sha256()
    parts = ["2", application_id, language, rule, sink["role"], sink["function"], sink["location"], sink["operation"], sink["path_role"], sink["input_part"], *sorted(set(signatures), key=lambda value: value.encode("utf-16-be", "surrogatepass"))]
    for part in parts:
        value = str(part).encode("utf-16", "surrogatepass").decode("utf-16", "replace").encode("utf-8")
        result.update(len(value).to_bytes(4, "big"))
        result.update(value)
    return "finding-" + result.hexdigest()


def trace_id(value, length):
    text = str(value or "").lower()
    return text if len(text) == length and re.fullmatch("[0-9a-f]+", text) and set(text) != {"0"} else ""


def component_reference(value=None, application_id=""):
    value = value or {}
    return {"status": value.get("status", "unresolved"), "sbom_id": value.get("sbom_id", ""), "revision": value.get("revision"), "release_id": value.get("release_id", ""), "application_id": value.get("application_id", application_id), "bom-ref": value.get("bom-ref", ""), "reason": value.get("reason", "" if value.get("status") == "resolved" else "component_not_resolved"), "observed_url": value.get("observed_url", ""), "query": value.get("query", "")}


def request_fields(value=None):
    return {"method": "", "route": "", "route_status": "unavailable", "status_code": None, "started_at": None, "ended_at": None, "framework": "", "transport": "", "error_type": "", **(value or {})}


def event_record(original, identity=None):
    event = dict(original)
    from .config import runtime_identity
    source = {"application_id": "", "instance_id": "", "service": {}, "code": {"repository": "", "commit": "", "build_id": "", "service_version": ""},
              "runtime": runtime_identity(), "identity_status": "incomplete", **(identity or {}), **(event.pop("identity", None) or {})}
    for field in ("application_id", "instance_id", "service", "code", "runtime", "identity_status"):
        if field not in event and field in source:
            event[field] = source[field]
    event["schema_version"] = SCHEMA_VERSION
    name = event.get("event_name")
    event["source"] = (
        "security_context_sbom"
        if name == "app-dependencies-loaded" or str(name).startswith("security.sbom.")
        else "security_context"
    )
    event.setdefault("observed_at", dt.datetime.now(dt.timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z"))
    if event.get("event_name") in ("security.dataflow.observed", "security.collection.incomplete"):
        event["request"] = request_fields(event.get("request"))
        event["trace_id"] = trace_id(event.get("trace_id"), 32)
        event["server_span_id"] = trace_id(event.get("server_span_id"), 16) if event["trace_id"] else ""
        event["current_span_id"] = trace_id(event.get("current_span_id"), 16) if event["trace_id"] else ""
        event["trace_flags"] = int(event.get("trace_flags") or 0) & 0xff if event["server_span_id"] else 0
        event["trace_availability"] = "not_guaranteed_by_trace_id"
    if event.get("event_name") == "security.dataflow.observed":
        event["component"] = component_reference(event.get("component"), event.get("application_id", ""))
        event.setdefault("stack", [])
    if name == "security.collection.incomplete":
        event["counts"] = {"objects": None, "nodes": None, "sources": None, "findings": None, "retained_bytes": None, **event.get("counts", {})}
        event.setdefault("collection_status", "unknown")
    if name == "app-dependencies-loaded":
        event.setdefault("status", "current")
        event.setdefault("dropped_observations", 0)
    if name == "security.sbom.health":
        defaults = {"sbom_id": "", "revision": 0, "release_id": "", "status": "initializing", "last_refresh_at": None, "last_failure_at": None,
                    "last_error_type": None, "current_components": None, "history_count": None, "completeness": "incomplete", "reasons": [], "dropped_observations": 0}
        for key, value in defaults.items():
            event.setdefault(key, value)
    return event

File: nodejs/src/schema.mjs
import { createHash } from 'node:crypto';
import { runtimeIdentity } from './config.mjs';

export const SCHEMA_VERSION = 2;
export const PRODUCT = 'SecurityContext';
export const FINGERPRINT_VERSION = 2;

export function sinkFields(rule, role, fn, location) {
  const aliases = { template: 'sql_template', shell: 'shell_script', shell_command: 'shell_script', argv: 'argument', ordinary_argument: 'argument', unknown_target: 'destination_unknown', request_path: 'path_or_query', request_query: 'path_or_query', path_query: 'path_or_query' };
  const sink = { function: fn, role: aliases[role] || role, location, operation: '', path_role: '', input_part: '' };
  if (rule === 'http_request_input') sink.input_part = role === 'request_path' ? 'path' : role === 'request_query' ? 'query' : 'path_or_query';
  if (rule === 'path_traversal') {
    sink.role = 'file_path';
    const name = fn.toLowerCase();
    let operation = 'unknown';
    if (/copy|\.cp(?:sync)?$|\.link(?:sync)?$/.test(name)) operation = 'copy';
    else if (/rename|\.move$/.test(name)) operation = 'rename';
    else if (/delete|unlink|rmdir|\.rm(?:sync)?$/.test(name)) operation = 'delete';
    else if (['read', 'write', 'delete', 'rename'].includes(role)) operation = role;
    else if (role === 'source' || /inputstream|reader|directorystream|\.read/.test(name)) operation = 'read';
    else if (role === 'target' || /outputstream|writer|\.write/.test(name)) operation = 'write';
    sink.operation = operation;
    sink.path_role = 'unknown';
    if (role === 'source' || role === 'read') sink.path_role = 'source';
    else if (['target', 'write', 'delete', 'destination_path'].includes(role)) sink.path_role = 'target';
    else if (role === 'file_path' && ['copy', 'rename', 'read'].includes(operation)) sink.path_role = 'source';
    else if (role === 'file_path' && ['write', 'delete'].includes(operation)) sink.path_role = 'target';
  }
  return sink;
}

export function findingFingerprint(applicationId, language, rule, sink, signatures) {
  const hash = createHash('sha256');
  const parts = ['2', applicationId, language, rule, sink.role, sink.function, sink.location, sink.operation, sink.path_role, sink.input_part, ...[...new Set(signatures)].sort()];
  for (const part of parts) {
    const value = Buffer.from(String(part), 'utf8');
    const length = Buffer.allocUnsafe(4);
    length.writeUInt32BE(value.length);
    hash.update(length).update(value);
  }
  return 'finding-' + hash.digest('hex');
}

export function traceId(value, length) {
  const text = String(value || '').toLowerCase();
  return text.length === length && /^[0-9a-f]+$/.test(text) && !/^0+$/.test(text) ? text : '';
}

export function componentReference(value = {}, applicationId = '') {
  value ||= {};
  return { status: value.status || 'unresolved', sbom_id: value.sbom_id || '', revision: value.revision ?? null, release_id: value.release_id || '', application_id: value.application_id || applicationId, 'bom-ref': value['bom-ref'] || '', reason: value.reason || (value.status === 'resolved' ? '' : 'component_not_resolved'), observed_url: value.observed_url || '', query: value.query || '' };
}

export function requestFields(value = {}) {
  return { method: '', route: '', route_status: 'unavailable', status_code: null, started_at: null, ended_at: null, framework: '', transport: '', error_type: '', ...value };
}

export function eventRecord(input, identity = {}) {
  const event = { ...input };
  const source = { application_id: '', instance_id: '', service: {}, code: { repository: '', commit: '', build_id: '', service_version: '' }, runtime: runtimeIdentity(), identity_status: 'incomplete', ...identity, ...(event.identity || {}) };
  delete event.identity;
  for (const field of ['application_id', 'instance_id', 'service', 'code', 'runtime', 'identity_status']) {
    if (event[field] === undefined && source[field] !== undefined) event[field] = source[field];
  }
  event.schema_version = SCHEMA_VERSION;
  event.source = event.event_name === 'app-dependencies-loaded' || String(event.event_name).startsWith('security.sbom.')
    ? 'security_context_sbom' : 'security_context';
  event.observed_at ||= new Date().toISOString();
  if (event.event_name === 'security.dataflow.observed' || event.event_name === 'security.collection.incomplete') {
    event.request = requestFields(event.request);
    event.trace_id = traceId(event.trace_id, 32);
    event.server_span_id = event.trace_id ? traceId(event.server_span_id, 16) : '';
    event.current_span_id = event.trace_id ? traceId(event.current_span_id, 16) : '';
    event.trace_flags = event.server_span_id ? (Number(event.trace_flags) || 0) & 0xff : 0;
    event.trace_availability = 'not_guaranteed_by_trace_id';
  }
  if (event.event_name === 'security.dataflow.observed') {
    event.component = componentReference(event.component, event.application_id);
    event.stack ||= [];
  }
  if (event.event_name === 'security.collection.incomplete') {
    event.counts = { objects: null, nodes: null, sources: null, findings: null, retained_bytes: null, ...event.counts };
    event.collection_status ||= 'unknown';
  }
  if (event.event_name === 'app-dependencies-loaded') {
    event.status ||= 'current';
    event.dropped_observations ??= 0;
  }
  if (event.event_name === 'security.sbom.health') {
    const defaults = { sbom_id: '', revision: 0, release_id: '', status: 'initializing', last_refresh_at: null, last_failure_at: null,
      last_error_type: null, current_components: null, history_count: null, completeness: 'incomplete', reasons: [], dropped_observations: 0 };
    for (const [key, value] of Object.entries(defaults)) if (event[key] === undefined) event[key] = value;
  }
  return event;
}
